### DISCUSSION TITLE
Phone mask input with multiple countries

### DISCUSSION THREAD

**Discussion by sitole** [mantinedev/mantine#3184] [Q&A] [answered]
Hi! We are using Mantine for our project and we love it! 🧡

Actually documentation is amazing and everything can be changed depend on our needs.
But for some time we are struggling with phone input implementation.

Actually we are testing `react-input-mask` package with Mantine `BaseInput`. It works great, we are ready to make some custom Yup validation (if string does not contains any `_` etc). But now there is challenge with multiple countries and phone formats.

Our current demo is something like this ⬇️ 
```typescript
import { InputBase } from '@mantine/core'
import InputMask from 'react-input-mask'
import { useState } from 'react'
import styled from 'styled-components'

 const countries = [
     {
         flag: '🇨🇿',
         code: 'CZ',
         name: 'Czechia',
         mask: '+420 999 999 999',
     },
     {
         flag: '🇺🇸',
         code: 'US',
         name: 'America',
         mask: '+1(999)999-9999',
     },
     {
         flag: '🇪🇸',
         code: 'ES',
         name: 'Spain',
         mask: '+34(999)999-999',
     },
 ]

 export const PhoneInput = (props: object) => {
     const [country, setCountry] = useState<typeof countries[0]>(countries[0])

     const handleCountryChange = (value: string) => {
         const findCountry = countries?.find((r) => r.code === value)

         if (findCountry) {
             setCountry(findCountry)
         }
     }

     return (
         <InputBase
             {...props}
             styles={{ icon: { pointerEvents: 'auto' } }}
             component={InputMask}
             mask={country.mask}
             alwaysShowMask={true}
             placeholder='Your phone'
             label='Your phone'
             iconWidth={40}
             icon={
                 <span>
                     <Selector
                         onChange={(e) => handleCountryChange(e.target.value)}
                     >
                         {countries.map((c) => (
                             <option key={c.code} value={c.code}>
                                 {c.flag} {c.name}
                             </option>
                         ))}
                     </Selector>

                     <span>{country.flag}</span>
                 </span>
             }
         />
     )
 }

 // Make select input invisible and act only in background
 const Selector = styled.select`
     position: absolute;

     top: 0;
     left: 0;
     height: 100%;
     width: 100%;

     z-index: 1;
     opacity: 0;
     cursor: pointer;

     border-top-width: 0;
     border-right-width: 0;
     border-bottom-width: 0;
     border-left-width: 0;
     border-top-style: initial;
     border-right-style: initial;
     border-bottom-style: initial;
     border-left-style: initial;
     border-top-color: initial;
     border-right-color: initial;
     border-bottom-color: initial;
     border-left-color: initial;
     border-image-source: initial;
     border-image-slice: initial;
     border-image-repeat: initial;
 `
```

this actually works great. We have hidden select input as input icon (actually country flag) and after you will select different country flag and input mask will be replaced.

Currently we are struggling with this points:
- How to support initial values (detect and call custom function and replace country flag/mask based on country prefix), in react hook forms there is `control` manager and you can listen/update all changes but is there same functionality in Mantine forms?
- Is there any standardized way how to create "input wrappers"? I want to make custom form but propagate all params from parent component `<PhoneInput label='My phone number' />` to `BaseInput`, its okay and it works but TypeScript types are 👹

Is there some ready-made solution or just example of Phone input based on Mantine forms? I would like to check it 🧡
Thanks for any help!

**Comment by rtivital** [comment, marked as answer]
> How to support initial values (detect and call custom function and replace country flag/mask based on country prefix)

You have access to `form.values` all the time, you can get value or subscribe to its changes with `useEffect`.

> Is there any standardized way how to create "input wrappers"?

You can explore source code and see how Mantine inputs are built with TypeScript. TextInput is a great starting point.
